linkerd/app/outbound/src/http/peer_proxy_errors.rs

```diff
@@ -0,0 +1,186 @@
+use http::{Request, Response};
+use linkerd_app_core::{
+    errors::L5D_PROXY_ERROR,
+    svc::{layer, NewService, Service},
+};
+use linkerd_proxy_http::ClientHandle;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// PeerProxyErrors will close the accepted connection if the response is from
+/// a peer proxy and the L5D_PROXY_ERROR header is set. This means the peer
+/// proxy encountered an inbound connection error with its application and
+/// therefore the accepted connection should be torn down.
+#[derive(Clone, Debug)]
+pub struct PeerProxyErrors<N> {
+    inner: N,
+}
+
+impl<N> PeerProxyErrors<N> {
+    fn new(inner: N) -> Self {
+        Self { inner }
+    }
+
+    pub fn layer() -> impl layer::Layer<N, Service = Self> + Clone + Copy {
+        layer::mk(Self::new)
+    }
+}
+
+impl<T, N> NewService<T> for PeerProxyErrors<N>
+where
+    N: NewService<T>,
+{
+    type Service = PeerProxyErrors<N::Service>;
+
+    fn new_service(&mut self, target: T) -> Self::Service {
+        let inner = self.inner.new_service(target);
+        PeerProxyErrors { inner }
+    }
+}
+
+impl<S, A, B> Service<Request<A>> for PeerProxyErrors<S>
+where
+    S: Service<Request<A>, Response = Response<B>>,
+    S::Response: Send,
+    S::Future: Send + 'static,
+    A: Send + 'static,
+    B: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<S::Response, S::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<A>) -> Self::Future {
+        let client = req.extensions().get::<ClientHandle>().cloned();
+        debug_assert!(client.is_some(), "Missing client handle");
+        let response = self.inner.call(req);
+        Box::pin(async move {
+            let response = response.await?;
+            if let Some(message) = response.headers().get(L5D_PROXY_ERROR) {
+                tracing::info!(?message, "response contained `{}` header", L5D_PROXY_ERROR);
+                // Gracefully teardown the accepted connection.
+                if let Some(ClientHandle { close, .. }) = client {
+                    tracing::info!("connection closed");
```

Comment:
this should probably not be info
```suggestion
                    tracing::trace!("connection closed");
```